### PR TITLE
Improving edit

### PR DESCRIPTION
**Pull request by joktz**
What I've done

1. Built a javascript controller to handle clicks on ingredient card content
2. Added some CSS to handle the input fields
3. Edited the ingredient controller to allow json edits through Javascript, so that the data could be edited without having to reload the page
4. Added some basic error messages within the cards (since the flash messages won't render, as the page isn't reloading)
5. Connected the edit controller and the datepicker controller to allow editing of the expiration date
6. Wrestled with that for hours, as flatpickr auto changes the items in the background. Built custom methods in the controller to handle that
7. Added validation in the ingredients model for expiration_date to prevent a blank entry on the edit
8. Fixed some bugs with things not rendering and added a guard against hitting the backspace while the flatpickr was open (which otherwise deleted the expiration date entirely)

Overall, I started with just the amount and it got a LOT more complicated the more I tried to do with it. Doing this without opening any forms was quite an undertaking, but I think it does make the page far more usable overall. Need to see how it work on a phone with a touchscreen--I'm worried that someone scrolling will accidentally click on all this stuff.

### CODE REVIEW

File: app/models/ingredient.rb
Comment (reidstivers) [inline app/models/ingredient.rb:1]: Weren't we discussing making expiration dates optional? I'm fine either way just wanted to bring that up for the case of non-perishables.

**Review by reidstivers — APPROVED** [review]
Great work as always, I think everything looks good we'll just have to test it on touch screen when it gets pushed to production. 

I left one comment about whether we want to add validation to expiration date, but if it needs to be that way so be it.

**Review by chiaki0607 — APPROVED** [review]
It looks great!!!!!

**Review by rachel-schlaifer — APPROVED** [review]
Great!

**Pull request merged**